From 4d20fcc2d0425d173864c6ab0e089f791ff528f1 Mon Sep 17 00:00:00 2001
From: Andrzej Kaczmarek <andrzej.kaczmarek@tieto.com>
Date: Wed, 22 Jan 2014 17:34:32 +0100
Subject: [PATCH] hid: Backport hid-generic driver

This patch provides provisional backport of hid-generic driver which
is required for Bluetooth HIDP to work with backports, since hidp
module has its HID driver removed.
---
 drivers/hid/Makefile            |  1 +
 drivers/hid/hid-core.c          |  2 +-
 drivers/hid/hid-generic.c       | 60 +++++++++++++++++++++++++++++++++++++++++
 include/linux/mod_devicetable.h |  1 +
 4 files changed, 63 insertions(+), 1 deletion(-)
 create mode 100644 drivers/hid/hid-generic.c

diff --git a/drivers/hid/Makefile b/drivers/hid/Makefile
index 9dca845..8fc013b 100644
--- a/drivers/hid/Makefile
+++ b/drivers/hid/Makefile
@@ -34,6 +34,7 @@ ifdef CONFIG_DEBUG_FS
 	hid-wiimote-y	+= hid-wiimote-debug.o
 endif
 
+obj-y				+= hid-generic.o
 obj-$(CONFIG_HID_A4TECH)	+= hid-a4tech.o
 obj-$(CONFIG_HID_ACRUX)		+= hid-axff.o
 obj-$(CONFIG_HID_APPLE)		+= hid-apple.o
diff --git a/drivers/hid/hid-core.c b/drivers/hid/hid-core.c
index 4da66b4..fa2f983 100644
--- a/drivers/hid/hid-core.c
+++ b/drivers/hid/hid-core.c
@@ -1158,7 +1158,7 @@ EXPORT_SYMBOL_GPL(hid_input_report);
 static bool hid_match_one_id(struct hid_device *hdev,
 		const struct hid_device_id *id)
 {
-	return id->bus == hdev->bus &&
+	return (id->bus == HID_BUS_ANY || id->bus == hdev->bus) &&
 		(id->vendor == HID_ANY_ID || id->vendor == hdev->vendor) &&
 		(id->product == HID_ANY_ID || id->product == hdev->product);
 }
diff --git a/drivers/hid/hid-generic.c b/drivers/hid/hid-generic.c
new file mode 100644
index 0000000..a49377f
--- /dev/null
+++ b/drivers/hid/hid-generic.c
@@ -0,0 +1,60 @@
+/*
+ *  HID support for Linux
+ *
+ *  Copyright (c) 1999 Andreas Gal
+ *  Copyright (c) 2000-2005 Vojtech Pavlik <vojtech@suse.cz>
+ *  Copyright (c) 2005 Michael Haboustak <mike-@cinci.rr.com> for Concept2, Inc
+ *  Copyright (c) 2007-2008 Oliver Neukum
+ *  Copyright (c) 2006-2012 Jiri Kosina
+ *  Copyright (c) 2012 Henrik Rydberg
+ *  Copyright (c) 2014 Andrzej Kaczmarek
+ */
+
+/*
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ */
+
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/kernel.h>
+#include <asm/unaligned.h>
+#include <asm/byteorder.h>
+
+#include <linux/hid.h>
+
+static const struct hid_device_id hid_generic_devices[] = {
+	{ HID_DEVICE(HID_BUS_ANY, HID_ANY_ID, HID_ANY_ID) },
+	{ }
+};
+MODULE_DEVICE_TABLE(hid, hid_generic_devices);
+
+static struct hid_driver hid_generic_driver = {
+	.name = "hid-generic",
+	.id_table = hid_generic_devices,
+};
+
+static int __init hid_generic_init2(void)
+{
+	int ret;
+
+	ret = hid_register_driver(&hid_generic_driver);
+	if (ret)
+		pr_err("can't register hid-generic driver\n");
+
+	return ret;
+}
+
+static void __exit hid_generic_exit2(void)
+{
+	hid_unregister_driver(&hid_generic_driver);
+}
+
+module_init(hid_generic_init2);
+module_exit(hid_generic_exit2);
+
+MODULE_AUTHOR("Henrik Rydberg");
+MODULE_DESCRIPTION("HID generic driver");
+MODULE_LICENSE("GPL");
diff --git a/include/linux/mod_devicetable.h b/include/linux/mod_devicetable.h
index 839767f..ec0878f 100644
--- a/include/linux/mod_devicetable.h
+++ b/include/linux/mod_devicetable.h
@@ -139,6 +139,7 @@ struct usb_device_id {
 #define USB_DEVICE_ID_MATCH_INT_NUMBER		0x0400
 
 #define HID_ANY_ID				(~0)
+#define HID_BUS_ANY				0xffff
 
 struct hid_device_id {
 	__u16 bus;
-- 
1.8.5.2

